File: tools/sefasi/sefasi-0.5/csrc/sefasi_sample_items.cc
/*
    Copyright (c) 2011-2012 Centre for Advanced Internet Architectures,
    Swinburne University of Technology.

    Author: Sebastian Zander (szander@swin.edu.au)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2 as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

    sefasi_sample_items.cc
    Sample item using deterministic sampling

    $Id: sefasi_sample_items.cc 864 2014-05-14 02:30:59Z szander $
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <getopt.h>
#include <string.h>
#include <math.h>
#include <err.h>
#include <errno.h>
#include <sysexits.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <iostream>
#include <iomanip>
#include <fstream>
#include <sstream>
#include <string>

#include "sefasi_common.h"
#include "murmur3.h"

using namespace std;



// all the hash names
typedef enum
{
	sha1 = 0,
	md5,
	murmur
} hash_names_t;	

// globals

// verbosity level
int verbose = 0;


inline __attribute__((always_inline)) uint32_t hash_val(hash_names_t hf, const void * key, int len, uint64_t salt)
{
	uint32_t ret;

	switch (hf) {
		case sha1:
			return 0;
		case md5:
			return 0;
		case murmur:
			// we just pass the salt as seed for performance reasons
			MurmurHash3_x86_32(key, len, (uint32_t) salt, &ret);
			//MurmurHash3_x64_128(key, len, salt, &ret);
			return ret;

		default:
			return 0;
	}
}


// print usage
void usage(char *name)
{

	cout << "Usage: " << name << " [OPTIONS] <-|item_list> " << endl;
	cout << "       -h                                      show usage" << endl;
	cout << "       -H <hash_function>         		hash function to be used ('sha1' or 'md5' or 'murmur', default 'murmur')" << endl;
	cout << "       -i <index>         			index of input file that contains IP addresses" << endl;
	cout << "       -r <sample_rate>         		sample rate that must be 1.0 or equal smaller 0.5 (default 1.0)" << endl;
	cout << "       -s <salt>         			random 'salt' that is xored with IP before hashing (default 0)" << endl;
	cout << "       -v       				enable verbose mode" << endl;
	cout << "       <-|item_list>      			file name to read item from or '-' to read from stdin" << endl;

}


int main(int argc, char **argv)
{
	int c;
	char *endptr;
	opterr = 0;
	FILE *item_file;

	hash_names_t hash = murmur;
	unsigned int idx = 0;
	//double sample_rate = 1.0;
	unsigned int sample_mod = 1;
	uint64_t salt = 0;

	//srand48(time(NULL));
	//srandom(time(NULL));

	while ((c = getopt(argc, argv, "hH:i:Ir:s:v")) != -1) {
		switch (c)
		{
			case 'H':
				if (strcasecmp(optarg, "sha1") == 0) {
					errx(EX_USAGE, "Error: SHA1 hash function not implemented yet");	
				} else if (strcasecmp(optarg, "md5") == 0) {
					errx(EX_USAGE, "Error: MD5 hash function not implemented yet");
				} else if (strcasecmp(optarg, "murmur") == 0) {
					hash = murmur;
				} else {
					errx(EX_USAGE, "Error: hash function '%s' not implemented", optarg);
				}
				break;

			case 'i':
				{
                                        unsigned long x = strtoul(optarg, &endptr, 10);
                                        if (*endptr) {
                                                errx(EX_USAGE, "Error: parsing index value '%s'", endptr);
                                        }

					idx = x;
                                }
                                break;

			case 'I':
				break;

			case 'r':
				{
					double x = strtod(optarg, &endptr);
					if (*endptr) {
						errx(EX_USAGE, "Error: parsing sample rate '%s'", endptr);
					}

					if (x < 0.0 || x > 1.0) {
						errx(EX_USAGE, "Error: sample rate must be between 0.0 and 1.0");
					}

					// XXX check on whether the inverse is an integer

					sample_mod = (unsigned int) round(1.0 / x);
					//sample_rate = x;

				}
				break;

			case 's':
                                {
                                        unsigned long long x = strtoull(optarg, &endptr, 10);
                                        if (*endptr) {
                                                errx(EX_USAGE, "Error: parsing salt value '%s'", endptr);
                                        }

                                        salt = x;
                                }
                                break;


			case 'h':
				usage(argv[0]);
				exit(0);
				break;

			case 'v':
				verbose = 1;
				break;

			case '?':
			default:
				usage(argv[0]);
				exit(1);
		}
	}

	// open file or stdin

	if (optind < argc) {
		if (strcmp(argv[optind], "-") == 0) {
			item_file = stdin;
		} else {
                	item_file = fopen(argv[optind], "r");
                	if (item_file == NULL) {
				errx(EX_USAGE, "Error: can't open item file '%s'", argv[optind]);
                	}
		}
        } else {
                usage(argv[0]);
                exit(1);
        }

	// processing loop

	char s1[MAX_ITEM_LEN+1];
	unsigned long line = 0;
	while (fscanf(item_file, "%"xstr(MAX_ITEM_LEN)"[^\n]\n", (char*)&s1) != EOF) {
		unsigned int i = 0;
		uint32_t h = 0;
		char *field = NULL;
		char *p = s1;

		while (i <= idx) {
			field = strsep(&p, " ");	
			if (field == NULL) {
				errx(EX_DATAERR, "Error: line %ld has fewer than %u elements", line, idx);	
			}
			i++;
		}
		// replace end of line 
		p = strstr(field, "\n");
               	if (p != NULL) {
                       	*p = '\0';
		}

		h = hash_val(hash, field, strlen(field), salt);
		if (h % sample_mod == 0) {
			printf("%s\n", field);
		}
		
		line++;
	}

	fclose(item_file);

	return 0;
}

File: tools/sefasi/sefasi-0.5/src/sefasi_sample_items.py
#!/usr/local/bin/python
# Copyright (c) 2013 Centre for Advanced Internet Architectures,
# Swinburne University of Technology. 
#
# Author: Sebastian Zander (szander@swin.edu.au)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as 
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
#
# Deterministically sample items 
#
# $Id: sefasi_sample_items.py 644 2013-08-15 05:27:44Z szander $

import os
import sys 
import getopt
import csv
from Crypto.Hash import SHA, MD5 
from Crypto.Util import number
import murmur

# column of input file that has IPs
idx = 0
# sample rate
srate = 1
# random salt value
salt = 0
use_salt = False
# hash function
hash_function_name = "murmur"
hash_function = -1
# be verbose
verbose = False


def enum(*sequential, **named):
	enums = dict(zip(sequential, range(len(sequential))), **named)
	return type('Enum', (), enums)


def usage():
	print "Usage: " + os.path.basename(sys.argv[0]) + " [-H <hash_function>] [-i <index>] [-r <sample_rate>] [-s <salt>] [-h] [-v] (-|<item_list>)\n" + \
		"\t-H <hash_function> \thash function to be used ('sha1' or 'md5' or 'murmur', default 'murmur')\n" + \
		"\t-i <index> \t\tindex of input file that contains IP addresses\n" + \
		"\t-r <sample_rate> \tsample rate that must be 1.0 or equal smaller 0.5 (default 1.0)\n" + \
		"\t-s <salt> \t\trandom 'salt' that is xored with IP before hashing (default 0)\n" + \
		"\t-h \t\t\tshow usage\n" + \
		"\t-v \t\t\tenable verbose mode"


def die(msg, show_usage=False):
        print msg
        if show_usage:
                usage()
        sys.exit(1)

# main

Hashes = enum('SHA1', 'MD5', 'MURMUR')

try:
	opts, args = getopt.getopt(sys.argv[1:], "H:i:r:s:hv")
except getopt.GetoptError, err:
	usage()
	die("Error: " + err.msg)	

for o, a in opts:
	if o == "-h":
		usage()
		sys.exit(0)
	elif o == "-H":
		hash_function_name = a
	elif o == "-i":
		idx = int(a) 
	elif o == "-r":
                srate = float(a)
	elif o == "-s":
		salt = long(a)
		use_salt = True
	elif o == "-v":
		verbose = True
	else:
		assert False, "unhandled option " + str(o)

if srate != 1.0 and (srate > 0.5 or srate < 0.0):
        die("Error: sample rate must be 1.0 or between 0.0 and 0.5")
if idx < 0:
	die("Error: IP column index must be 0 or greater")

hash_function = Hashes.MURMUR
if cmp(hash_function_name, "sha1") == 0 :
	hash_function = Hashes.SHA1
elif cmp(hash_function_name, "md5") == 0 :
	hash_function = Hashes.MD5
elif cmp(hash_function_name, "murmur") == 0 :
        hash_function = Hashes.MURMUR
else:
	die("Error: hash function must be 'sha1' or 'md5' or 'murmur'")
	
if (cmp(sys.argv[-1], "-") == 0):
        reader = csv.reader(sys.stdin, delimiter=' ')
else:
        reader = csv.reader(open(sys.argv[-1],"rb"), delimiter=' ')

mod = int(round(1.0 / srate, 0))

if verbose:
	esrate = 1.0 / float(mod)
	print "Info: effective sample rate is", esrate
	if salt > 0:
		print "Info: using salt value", salt

for row in reader:

	try:
		if not use_salt:
			b = row[idx] 
		else:
			b = row[idx] + str(salt)
	except IndexError:
		die("Error: input file has less than index columns")
	
	if hash_function == Hashes.MURMUR:
	        h = murmur.string_hash(b)

		if int(h) % mod == 0 :
                        print ' '.join(row)
	elif hash_function == Hashes.SHA1:
		h = SHA.new(b)		

		if number.bytes_to_long(h.digest()) % mod == 0 :
                        print ' '.join(row)
	else:
		h = MD5.new(b)

		if number.bytes_to_long(h.digest()) % mod == 0 :
			print ' '.join(row)

